README
Mirror of aufs2-standalone

ChangeLog
commit ba8f448c3c5da969b4b51503a7ac0015785bac6c
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Sat Apr 3 16:37:36 2010 +0900

    ifdef __KERNEL__ for kernel.h
    
    For those who doesn't run "make headers_install" and build the module
    out of kernel source tree.
    Since linux-2.6.32, include/linux/kernel.h warns about being included
    from userspace.
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 8d462ecd47082b966339a0620b25ada5b9800bfc
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Thu Apr 1 22:06:18 2010 +0900

    aufs: bugfix, error from kset_create_and_add()
    
    It can be NULL.
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 5502e7b6925ae1c928ccc8be272bb988575c24f1
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Thu Apr 1 14:46:11 2010 +0900

    aufs: tiny, revise including header files
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 0e40748ad3c6200339212eb2626c2ab50c189585
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Wed Mar 31 09:09:08 2010 +0900

    aufs: bugfix, missing unlock in an error path
    
    Forgot unlocking after an error in au_lock_and_icpup().
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 9df16a8b1ccc7240d734ab6b50943063d67e311a
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Wed Mar 31 09:07:37 2010 +0900

    aufs: bugfix, support updating a removed dir
    
    In au_cpup_wh(), test the target is a dir or not.
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 4b1eb74536a0a56f28e3dc2652153de8fab54806
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Wed Mar 31 03:48:24 2010 +0900

    aufs: minor optimization, remove unnecessary au_set_h_[id]ptr()
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 9a8dac1e0e15535ba87b788c98af51d910133695
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 22:06:04 2010 +0900

    aufs: tiny, new conf HFSNOTIFY and HINOTIFY
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 568ec06d9f85f4fb633c5626f89c7f4ae1db5f3c
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 09:09:49 2010 +0900

    aufs: tiny, fix the comment style
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 48b705d5be75fa47d68a5d9502233b925e044e09
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 08:11:39 2010 +0900

    aufs: support FSNOTIFY, export functions
    
    Since FSNOTIFY is introduced in linux-2.6.31 and INOTIFY is on the list
    Documentation/feature-removal-schedule.txt, aufs's udba=inotify will be
    unavailable in the near future.
    
    Export FSNOTIFY functions for modules.
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 8842b479cb1afb25045554ea19f4ce24a42cd77c
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 04:32:04 2010 +0900

    aufs: support FSNOTIFY
    
    Since FSNOTIFY is introduced in linux-2.6.31 and INOTIFY is on the list
    Documentation/feature-removal-schedule.txt, aufs's udba=inotify will be
    unavailable in the near future.
    
    This commit introducces a new configuration AUFS_HFSNOTIFY which behaves
    as a toggle to AUFS_HINOTIFY.
    - introduce a new file hfsnotify.c
    - the common routine hnotify.c follows the symbols of FSNOTIFY, so
      au_hin_conv_mask() is introduced in hinotify.c. this new function
      convers simbols of INOTIFY into FSNOTIFY's.
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 4faae8f0a041f55fb2d00d6ad82486cf0bbeb38c
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 02:44:08 2010 +0900

    aufs: bugfix, stop setting the inotify watch twice for the same lower inode
    
    When the upper dir was removed by udba, but the same named dir left. In
    this case, aufs assignes a new inode number and set the monitor again.
    For the lower dir, the old monitnor is still left.
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 5ba60b523ebf6135bbc08f4bf1aff853e3e36807
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 02:13:28 2010 +0900

    aufs: tiny, remove the meaningles parameter
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit 067f28f06291a5f62b88b59d02959d03ed995c78
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 01:37:04 2010 +0900

    aufs: tiny, abstract hnotify.c, restore inotify specific code
    
    Since FSNOTIFY is introduced in linux-2.6.31 and INOTIFY is on the list
    Documentation/feature-removal-schedule.txt, aufs's udba=inotify will be
    unavailable in the near future.
    
    This commit restores inotify specific code to hinotify.c
    - introduce struct au_hnotify_op.
      + implement au_hin_{alloc,free,ctl,init,fin} in hinotify.c.
      + make au_hn_{alloc,free,ctl,init,fin} in hnotify.c abstract.
    - extract the core of aufs_inotify() and create a new function
      au_hnotify() in hnotify.c.
    - move the temporary declarations from inode.h to hnotify.c.
    - rename au_hnotify() to au_hn_bh().
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit b805c550d7841700037ac4057b462ba32bce5a42
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 00:04:44 2010 +0900

    replace AUFS_HINOTIFY by AUFS_HNOTIFY
    
    Signed-off-by: J. R. Okajima <hooanon05@yahoo.co.jp>

commit db1bcb4505efab21d60286f49bb7b1003b9636a3
Author: J. R. Okajima <hooanon05@yahoo.co.jp>
Date:   Fri Mar 26 00:01:42 2010 +0900

    aufs: tiny, tmp, abstract fs/aufs/hinotify.c, rename 'hin' prefix
    
    Since FSNOTIFY is introduced in linux-2.6.31 and INOTIFY is on the list
    Documentation/feature-removal-schedule.txt, aufs's udba=inotify will be
    unavailable in the near future.
    
    This commit makes an abstraction for UDBA notification.
    - since the prefix 'hin' means hidden inotify, replace most of them by

fs/aufs/sysfs.c
/*
 * Copyright (C) 2005-2010 Junjiro R. Okajima
 *
 * This program, aufs is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/*
 * sysfs interface
 */

#include <linux/fs.h>
#include <linux/module.h>
#include <linux/seq_file.h>
#include <linux/sysfs.h>
#include "aufs.h"

#ifdef CONFIG_AUFS_FS_MODULE
/* this entry violates the "one line per file" policy of sysfs */
static ssize_t config_show(struct kobject *kobj, struct kobj_attribute *attr,
			   char *buf)
{
	ssize_t err;
	static char *conf =
/* this file is generated at compiling */
#include "conf.str"
		;

	err = snprintf(buf, PAGE_SIZE, conf);
	if (unlikely(err >= PAGE_SIZE))
		err = -EFBIG;
	return err;
}

static struct kobj_attribute au_config_attr = __ATTR_RO(config);
#endif

static struct attribute *au_attr[] = {
#ifdef CONFIG_AUFS_FS_MODULE
	&au_config_attr.attr,
#endif
	NULL,	/* need to NULL terminate the list of attributes */
};

static struct attribute_group sysaufs_attr_group_body = {
	.attrs = au_attr
};

struct attribute_group *sysaufs_attr_group = &sysaufs_attr_group_body;

/* ---------------------------------------------------------------------- */

int sysaufs_si_xi_path(struct seq_file *seq, struct super_block *sb)
{
	int err;

	SiMustAnyLock(sb);

	err = 0;
	if (au_opt_test(au_mntflags(sb), XINO)) {
		err = au_xino_path(seq, au_sbi(sb)->si_xib);
		seq_putc(seq, '\n');
	}
	return err;
}

/*
 * the lifetime of branch is independent from the entry under sysfs.
 * sysfs handles the lifetime of the entry, and never call ->show() after it is
 * unlinked.
 */
static int sysaufs_si_br(struct seq_file *seq, struct super_block *sb,
			 aufs_bindex_t bindex)
{
	struct path path;
	struct dentry *root;
	struct au_branch *br;

	AuDbg("b%d\n", bindex);

	root = sb->s_root;
	di_read_lock_parent(root, !AuLock_IR);
	br = au_sbr(sb, bindex);
	path.mnt = br->br_mnt;
	path.dentry = au_h_dptr(root, bindex);
	au_seq_path(seq, &path);
	di_read_unlock(root, !AuLock_IR);
	seq_printf(seq, "=%s\n", au_optstr_br_perm(br->br_perm));
	return 0;
}

/* ---------------------------------------------------------------------- */

static struct seq_file *au_seq(char *p, ssize_t len)
{
	struct seq_file *seq;

	seq = kzalloc(sizeof(*seq), GFP_NOFS);
	if (seq) {
		/* mutex_init(&seq.lock); */
		seq->buf = p;
		seq->size = len;
		return seq; /* success */
	}

	seq = ERR_PTR(-ENOMEM);
	return seq;
}

#define SysaufsBr_PREFIX "br"

/* todo: file size may exceed PAGE_SIZE */
ssize_t sysaufs_si_show(struct kobject *kobj, struct attribute *attr,
			char *buf)
{
	ssize_t err;
	long l;
	aufs_bindex_t bend;
	struct au_sbinfo *sbinfo;
	struct super_block *sb;
	struct seq_file *seq;
	char *name;
	struct attribute **cattr;

	sbinfo = container_of(kobj, struct au_sbinfo, si_kobj);
	sb = sbinfo->si_sb;
	si_noflush_read_lock(sb);

	seq = au_seq(buf, PAGE_SIZE);
	err = PTR_ERR(seq);
	if (IS_ERR(seq))
		goto out;

	name = (void *)attr->name;
	cattr = sysaufs_si_attrs;
	while (*cattr) {
		if (!strcmp(name, (*cattr)->name)) {
			err = container_of(*cattr, struct sysaufs_si_attr, attr)
				->show(seq, sb);
			goto out_seq;
		}
		cattr++;
	}

	bend = au_sbend(sb);
	if (!strncmp(name, SysaufsBr_PREFIX, sizeof(SysaufsBr_PREFIX) - 1)) {
		name += sizeof(SysaufsBr_PREFIX) - 1;
		err = strict_strtol(name, 10, &l);
		if (!err) {
			if (l <= bend)
				err = sysaufs_si_br(seq, sb, (aufs_bindex_t)l);
			else
				err = -ENOENT;
		}
		goto out_seq;
	}
	BUG();

 out_seq:
	if (!err) {
		err = seq->count;
		/* sysfs limit */
		if (unlikely(err == PAGE_SIZE))
			err = -EFBIG;
	}
	kfree(seq);
 out:
	si_read_unlock(sb);
	return err;
}

/* ---------------------------------------------------------------------- */

void sysaufs_br_init(struct au_branch *br)
{
	br->br_attr.name = br->br_name;
	br->br_attr.mode = S_IRUGO;
	br->br_attr.owner = THIS_MODULE;
}

void sysaufs_brs_del(struct super_block *sb, aufs_bindex_t bindex)
{
	struct au_branch *br;
	struct kobject *kobj;
	aufs_bindex_t bend;

	dbgaufs_brs_del(sb, bindex);

	if (!sysaufs_brs)
		return;

	kobj = &au_sbi(sb)->si_kobj;
	bend = au_sbend(sb);
	for (; bindex <= bend; bindex++) {
		br = au_sbr(sb, bindex);
		sysfs_remove_file(kobj, &br->br_attr);
	}
}

void sysaufs_brs_add(struct super_block *sb, aufs_bindex_t bindex)
{
	int err;
	aufs_bindex_t bend;
	struct kobject *kobj;
	struct au_branch *br;

	dbgaufs_brs_add(sb, bindex);

	if (!sysaufs_brs)
		return;

	kobj = &au_sbi(sb)->si_kobj;
	bend = au_sbend(sb);
	for (; bindex <= bend; bindex++) {
		br = au_sbr(sb, bindex);
		snprintf(br->br_name, sizeof(br->br_name), SysaufsBr_PREFIX
			 "%d", bindex);
		err = sysfs_create_file(kobj, &br->br_attr);
		if (unlikely(err))
			pr_warning("failed %s under sysfs(%d)\n",
				   br->br_name, err);
	}
}


include/linux/aufs_type.h
/*
 * Copyright (C) 2005-2010 Junjiro R. Okajima
 *
 * This program, aufs is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifndef __AUFS_TYPE_H__
#define __AUFS_TYPE_H__

#include <linux/ioctl.h>
/* for those who didn't "make headers_install" */
#ifdef __KERNEL__
#include <linux/kernel.h>
#endif
#include <linux/limits.h>
#include <linux/types.h>

#define AUFS_VERSION	"2-standalone.tree-20100405"

/* todo? move this to linux-2.6.19/include/magic.h */
#define AUFS_SUPER_MAGIC	('a' << 24 | 'u' << 16 | 'f' << 8 | 's')

/* ---------------------------------------------------------------------- */

#ifdef CONFIG_AUFS_BRANCH_MAX_127
typedef __s8 aufs_bindex_t;
#define AUFS_BRANCH_MAX 127
#else
typedef __s16 aufs_bindex_t;
#ifdef CONFIG_AUFS_BRANCH_MAX_511
#define AUFS_BRANCH_MAX 511
#elif defined(CONFIG_AUFS_BRANCH_MAX_1023)
#define AUFS_BRANCH_MAX 1023
#elif defined(CONFIG_AUFS_BRANCH_MAX_32767)
#define AUFS_BRANCH_MAX 32767
#endif
#endif

#ifdef __KERNEL__
#ifndef AUFS_BRANCH_MAX
#error unknown CONFIG_AUFS_BRANCH_MAX value
#endif
#endif /* __KERNEL__ */

/* ---------------------------------------------------------------------- */

#define AUFS_NAME		"aufs"
#define AUFS_FSTYPE		AUFS_NAME

#define AUFS_ROOT_INO		2
#define AUFS_FIRST_INO		11

#define AUFS_WH_PFX		".wh."
#define AUFS_WH_PFX_LEN		((int)sizeof(AUFS_WH_PFX) - 1)
#define AUFS_WH_TMP_LEN		4
/* a limit for rmdir/rename a dir */
#define AUFS_MAX_NAMELEN	(NAME_MAX \
				- AUFS_WH_PFX_LEN * 2	/* doubly whiteouted */\
				- 1			/* dot */\
				- AUFS_WH_TMP_LEN)	/* hex */
#define AUFS_XINO_FNAME		"." AUFS_NAME ".xino"
#define AUFS_XINO_DEFPATH	"/tmp/" AUFS_XINO_FNAME
#define AUFS_XINO_TRUNC_INIT	64 /* blocks */
#define AUFS_XINO_TRUNC_STEP	4  /* blocks */
#define AUFS_DIRWH_DEF		3
#define AUFS_RDCACHE_DEF	10 /* seconds */
#define AUFS_RDBLK_DEF		512 /* bytes */
#define AUFS_RDHASH_DEF		32
#define AUFS_WKQ_NAME		AUFS_NAME "d"
#define AUFS_MFS_SECOND_DEF	30 /* seconds */
#define AUFS_PLINK_WARN		100 /* number of plinks */

#define AUFS_DIROPQ_NAME	AUFS_WH_PFX ".opq" /* whiteouted doubly */
#define AUFS_WH_DIROPQ		AUFS_WH_PFX AUFS_DIROPQ_NAME

#define AUFS_BASE_NAME		AUFS_WH_PFX AUFS_NAME
#define AUFS_PLINKDIR_NAME	AUFS_WH_PFX "plnk"
#define AUFS_ORPHDIR_NAME	AUFS_WH_PFX "orph"

/* doubly whiteouted */
#define AUFS_WH_BASE		AUFS_WH_PFX AUFS_BASE_NAME
#define AUFS_WH_PLINKDIR	AUFS_WH_PFX AUFS_PLINKDIR_NAME
#define AUFS_WH_ORPHDIR		AUFS_WH_PFX AUFS_ORPHDIR_NAME

/* branch permission */
#define AUFS_BRPERM_RW		"rw"
#define AUFS_BRPERM_RO		"ro"
#define AUFS_BRPERM_RR		"rr"
#define AUFS_BRPERM_WH		"wh"
#define AUFS_BRPERM_NLWH	"nolwh"
#define AUFS_BRPERM_ROWH	AUFS_BRPERM_RO "+" AUFS_BRPERM_WH
#define AUFS_BRPERM_RRWH	AUFS_BRPERM_RR "+" AUFS_BRPERM_WH
#define AUFS_BRPERM_RWNLWH	AUFS_BRPERM_RW "+" AUFS_BRPERM_NLWH

/* ---------------------------------------------------------------------- */

/* ioctl */
enum {
	AuCtl_PLINK_MAINT,
	AuCtl_PLINK_CLEAN,

	/* readdir in userspace */
	AuCtl_RDU,
	AuCtl_RDU_INO,

	/* pathconf wrapper */
	AuCtl_WBR_FD
};

/* borrowed from linux/include/linux/kernel.h */
#ifndef ALIGN
#define ALIGN(x, a)		__ALIGN_MASK(x, (typeof(x))(a)-1)
#define __ALIGN_MASK(x, mask)	(((x)+(mask))&~(mask))
#endif

/* borrowed from linux/include/linux/compiler-gcc3.h */
#ifndef __aligned
#define __aligned(x)			__attribute__((aligned(x)))
#define __packed			__attribute__((packed))
#endif

struct au_rdu_cookie {
	__u64		h_pos;
	__s16		bindex;
	__u8		flags;
	__u8		pad;
	__u32		generation;
} __aligned(8);

struct au_rdu_ent {
	__u64		ino;
	__s16		bindex;
	__u8		type;
	__u8		nlen;
	__u8		wh;
	char		name[0];
} __aligned(8);

static inline int au_rdu_len(int nlen)
{
	/* include the terminating NULL */
	return ALIGN(sizeof(struct au_rdu_ent) + nlen + 1,
		     sizeof(__u64));
}

union au_rdu_ent_ul {
	struct au_rdu_ent __user	*e;
	unsigned long			ul;
};

enum {
	AufsCtlRduV_SZ,
	AufsCtlRduV_SZ_PTR,
	AufsCtlRduV_End
};

struct aufs_rdu {
	/* input */
	union {
		__u64		sz;	/* AuCtl_RDU */
		__u64		nent;	/* AuCtl_RDU_INO */
	};
	union au_rdu_ent_ul	ent;
	__u16			verify[AufsCtlRduV_End];

	/* input/output */
	__u32			blk;

	/* output */
	union au_rdu_ent_ul	tail;
	/* number of entries which were added in a single call */
	__u64			rent;
	__u8			full;
	__u8			shwh;

	struct au_rdu_cookie	cookie;
} __aligned(8);

#define AuCtlType		'A'
#define AUFS_CTL_PLINK_MAINT	_IO(AuCtlType, AuCtl_PLINK_MAINT)
#define AUFS_CTL_PLINK_CLEAN	_IO(AuCtlType, AuCtl_PLINK_CLEAN)
#define AUFS_CTL_RDU		_IOWR(AuCtlType, AuCtl_RDU, struct aufs_rdu)
#define AUFS_CTL_RDU_INO	_IOWR(AuCtlType, AuCtl_RDU_INO, struct aufs_rdu)
#define AUFS_CTL_WBR_FD		_IO(AuCtlType, AuCtl_WBR_FD)

#endif /* __AUFS_TYPE_H__ */
